pandas/tests/reshape/merge/test_merge.py

```diff
@@ -2598,3 +2598,37 @@ def test_merge_outer_with_NaN(dtype):
         dtype=dtype,
     )
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("string_dtype", tm.STRING_DTYPES)
+def test_merge_ea_with_string(join_type, string_dtype):
+    # GH 43734 Avoid the use of `assign` with multi-index
+    df1 = DataFrame(
+        data={
+            ("lvl0", "lvl1-a"): ["1", "2", "3", "4", None],
+            ("lvl0", "lvl1-b"): ["4", "5", "6", "7", "8"],
+        },
+        dtype=pd.StringDtype(),
+    )
+    df2 = DataFrame(
+        data={
+            ("lvl0", "lvl1-a"): ["1", "2", "3", pd.NA, "5"],
+            ("lvl0", "lvl1-c"): ["7", "8", "9", pd.NA, "11"],
+        },
+        dtype=string_dtype,
+    )
+    merged = merge(left=df1, right=df2, on=[("lvl0", "lvl1-a")], how=join_type)
+
+    # No change in df1 and df2 types
+    assert (df1.dtypes == pd.StringDtype()).all()
+    assert (df2.dtypes == np.dtype("O")).all()
+
+    # Check the expected types for the merged data frame
+    result = merged.dtypes.sort_index()
```

Comment:
Don't use sort, just define expected correctly

Reply:
I picked this way of doing in the class `TestMergeCategorical` (in the same file). Done.